Add skip_verify for archive.extracted

Ch3LL opened the pull request:
### What does this PR do?

Adds the option skip_verify for archive.extracted so a user has an option to not specify a source_hash.
### What issues does this PR fix or reference?
#34515
### Previous Behavior
1. given an sls file like so:

```
vagrant-sdk:
  archive:
    - extracted
    - name: /opt/virtualbox/5.0.24
    - source: http://download.virtualbox.org/virtualbox/5.0.24/VirtualBoxSDK-5.0.24-108355.zip
    - archive_format: zip
    - user: ch3ll
    - group: staff
    - trim_output: 5
    - if_missing: /opt/virtualbox/5.0.24/sdk/installer
    - skip_verify: True
```

When attempting to call would receive this error:

```
 ch3ll@localhost  ~/git/salt   tagv2016.3.1  sudo salt-call --local state.sls issues.34515
[ERROR   ] Unable to determine upstream hash of source file http://download.virtualbox.org/virtualbox/5.0.24/VirtualBoxSDK-5.0.24-108355.zip
[ERROR   ] Unable to determine upstream hash of source file http://download.virtualbox.org/virtualbox/5.0.24/VirtualBoxSDK-5.0.24-108355.zip
local:
----------
          ID: vagrant-sdk
    Function: archive.extracted
        Name: /opt/virtualbox/5.0.24
      Result: False
     Comment: Unable to determine upstream hash of source file http://download.virtualbox.org/virtualbox/5.0.24/VirtualBoxSDK-5.0.24-108355.zip
     Started: 14:45:37.122698
    Duration: 85.373 ms
     Changes:   
    Warnings: 'skip_verify' is an invalid keyword argument for
              'archive.extracted'. If you were trying to pass additional data to
              be used in a template context, please populate 'context' with
              'key: value' pairs. Your approach will work until Salt Carbon is
              out. Please update your state files.

Summary for local
------------
Succeeded: 0
Failed:    1
Warnings:  1
------------
Total states run:     1
```
### New Behavior

```
 ch3ll@localhost  ~/git/salt   tagv2016.3.1  sudo salt-call --local state.sls issues.34515
[ERROR   ] Group staff is not available
local:
----------
          ID: vagrant-sdk
    Function: archive.extracted
        Name: /opt/virtualbox/5.0.24
      Result: True
     Comment: http://download.virtualbox.org/virtualbox/5.0.24/VirtualBoxSDK-5.0.24-108355.zip extracted to /opt/virtualbox/5.0.24/
     Started: 14:46:15.630102
    Duration: 777.949 ms
     Changes:   
              ----------
              directories_created:
                  - /opt/virtualbox/5.0.24/
              extracted_files:
                  - sdk/
                  - sdk/bindings/
                  - sdk/bindings/VirtualBox.xidl
                  - sdk/bindings/webservice/
                  - sdk/bindings/webservice/java/
                  - List trimmed after 5 files.

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
```
### Tests written?

Yes

ping @terminalmage would you mind reviewing this small change since you added skip_verify. Just want to make sure I didn't forget about something. thanks

Ch3LL committed:
add skip_verify option to archive.extracted

cachedout commented on salt/states/archive.py:
```diff
@@ -174,6 +175,13 @@ def extracted(name,
 
         .. versionadded:: 2016.3.0
 
+   skip_verify:False
```
Putting the default arg here is somewhat inconsistent compared to the other docs. Maybe move this to the body of the description?

Ch3LL replied:
doh missed a space

cachedout commented on tests/integration/states/archive.py:
```diff
@@ -0,0 +1,70 @@
+import salt.utils
+
+STATE_DIR = os.path.join(integration.FILES, 'file', 'base')
+ARCHIVE_DIR = '/tmp/archive/'
```
This is going to break on Windows. We need to use attributes of the integration module to construct a platform-agnostic directory.

Ch3LL replied:
ahh good catch. will add the windows paths as well. thanks

cachedout commented on tests/integration/states/archive.py:
```diff
@@ -0,0 +1,70 @@
+
+STATE_DIR = os.path.join(integration.FILES, 'file', 'base')
+ARCHIVE_DIR = '/tmp/archive/'
+ARCHIVE_TAR_SOURCE = 'https://github.com/downloads/Graylog2/'\
```
I think this is backed by `file.managed` which means we can use a `salt://` URI. I'd much rather have the master deliver this file then depend on an external source. Perhaps test with that approach? (I could be totally wrong here. Going on memory.)

Ch3LL replied:
You are right you can use `salt://` with archive, but when I went to change the test to using `salt:// `instead of `http` I found out that we do not check the `source_hash` on purpose if the file is on the salt master [here](https://github.com/saltstack/salt/blob/2016.3/salt/modules/file.py#L4436) , which makes the test to ensure the source_hash is checked correctly obsolete. So I do not see a way around it for that test. I can add  a tar.gz to the test suite for the other test and any other tests added in the future. are you okay with that? Or we could add an argument to check for source on the salt master in the code.

cachedout replied:
I'm fine with adding a `.tar.gz` to the test suite.

cachedout commented:
This is excellent work. A few small comments in-line.

thatch45 commented:
@Ch3LL is this ready for merge? What more do we need to do here?

Ch3LL committed:
add windows path and add custom tar

Ch3LL commented:
Apologies on the wait. I added the changes. There is still one test that is depended on an external source due to the fact that source_hash is not checked with `salt://` sources. Also added the windows paths. Will check up on this on monday to make sure the tests passed.

Ch3LL commented:
sorry, but I just rethought this implementation with more detail.  I believe the best way to approach this is to use external sources for both tests. because even the `test_archive_extracted_skip_verify` is not truly testing the `skip_verify` function,because salt:// sources never check the source_hash anways and therefore skip_verify becomes obsolete. I understand the reasons behind attempting to make the test local to the test suite but I do not see a way around this to ensure `skip_verify` source is tested correctly without using external sources. Are you okay with changing it back to external sources  @cachedout or @thatch45 ?

UtahDave commented:
@Ch3ll maybe we could put the tar file on a web server we control so it doesn't get moved out from under us. We could create a subdirectory on repo.saltstack.com, if that is convenient

Ch3LL commented:
@UtahDave yeah I love that idea. Would you mind helping me out with that? repo.saltstack.com works great for me.

cachedout commented:
IMHO, we should build a virtual server in our data center just for this purpose.

cc: @cro @jtand

s0undt3ch commented:
Would starting a local webserver using plain python stdlib to serve the test files serve the purpose here?

s0undt3ch commented:
We already depend on tornado, use its static web handler to serve the file(s)...

cachedout commented:
I love @s0undt3ch 's idea!

Ch3LL commented:
Perfect I will look into implementing this during setup of the test.

vernondcole commented:
As requested, I confirm that this pull request solves my issue #34515.
Tested on a clean install of Ubuntu 16.04.

Ch3LL committed:
add tornado web app to serve up static file for test

Ch3LL commented:
ping @cachedout I don't believe those test failures are related. So I believe my PR is ready for review. I added the web server using tornado. Please let me know if there is anyway to improve this thanks

cachedout commented:
@Ch3LL These tests are failing on a number of platforms. Could you please take a look and either disable them or fix them? Thanks.

Ch3LL commented:
should be fixed now
